alchemist/alchemist-swingui/src/main/java/it/unibo/alchemist/boundary/gui/effects/DrawLayers.java

```diff
@@ -0,0 +1,292 @@
+ */
+public abstract class DrawLayers implements Effect {
+
+    /**
+     */
+    protected static final int MAX_COLOUR_VALUE = 255;
+    /**
+     */
+    protected static final int INITIAL_ALPHA_DIVIDER = 2;
+    /**
+     */
+    protected static final Logger L = LoggerFactory.getLogger(DrawShape.class);
+    private static final long serialVersionUID = 1L;
+    @ExportForGUI(nameToExport = "Draw only layer containing a molecule")
+    private boolean layerFilter;
+    @ExportForGUI(nameToExport = "Incarnation to use")
+    private CollectionWithCurrentElement<String> curIncarnation;
+    @ExportForGUI(nameToExport = "Molecule")
+    private String molString = "";
+    @ExportForGUI(nameToExport = "A")
+    private RangedInteger alpha = new RangedInteger(0, MAX_COLOUR_VALUE, MAX_COLOUR_VALUE / INITIAL_ALPHA_DIVIDER);
+    @ExportForGUI(nameToExport = "R")
+    private RangedInteger red = new RangedInteger(0, MAX_COLOUR_VALUE);
+    @ExportForGUI(nameToExport = "G")
+    private RangedInteger green = new RangedInteger(0, MAX_COLOUR_VALUE);
+    @ExportForGUI(nameToExport = "B")
+    private RangedInteger blue = new RangedInteger(0, MAX_COLOUR_VALUE, MAX_COLOUR_VALUE);
+    private Color colorCache = Color.BLUE;
+    @Nullable
+    @SuppressFBWarnings("SE_TRANSIENT_FIELD_NOT_RESTORED")
+    private transient Molecule molecule;
+    @Nullable
+    @SuppressFBWarnings("SE_TRANSIENT_FIELD_NOT_RESTORED")
+    private transient Object molStringCached;
+    @SuppressFBWarnings(value = "SE_TRANSIENT_FIELD_NOT_RESTORED", justification = "If null, it gets reinitialized anyway if needed")
+    private transient CollectionWithCurrentElement<String> prevIncarnation;
+    private transient Incarnation<?, ?> incarnation;
+    private transient Optional<Node> markerNode = Optional.empty();
+
+    /**
+     */
+    public DrawLayers() {
+        final Set<String> availableIncarnations = SupportedIncarnations.getAvailableIncarnations();
+        if (availableIncarnations.isEmpty()) {
+            throw new IllegalStateException(getClass().getSimpleName() + " can't work if no incarnation is available.");
+        }
+        curIncarnation = new ImmutableCollectionWithCurrentElement<>(availableIncarnations, availableIncarnations.stream().findAny().get());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @SuppressWarnings({"PMD.CompareObjectsWithEquals", "unchecked"})
+    @SuppressFBWarnings("ES_COMPARING_STRINGS_WITH_EQ")
+    @Override
+    public <T, P extends Position2D<P>> void apply(final Graphics2D g, final Node<T> n, final Environment<T, P> env, final IWormhole2D<P> wormhole) {
+        // if marker node is no longer in the environment or it is no longer displayed, we need to change it
+        if (markerNode.isPresent() && (!env.getNodes().contains(markerNode.get())
+                                        || !wormhole.isInsideView(wormhole.getViewPoint(env.getPosition((Node<T>) markerNode.get()))))) {
```

Comment:
Why so indented?

Reply:
```java
// When I need to put something like this on multiple lines
if (condition1 && (condition2 || condition3)) {}
// I do something like this
if (condition1 && (condition2 
                   || condition3) {}
```
isn't it more clear that `condition3` is inside brackets?

If not, I will change it

Reply:
No, it's horrible :)
Break on logic disjunction, preferably

```java
if (condition1
        && (condition2 || condition3) {}
```

if it gets too long, break using variables

Reply:
Ok, fixed